### PR TITLE
Emotes

### LINKED ISSUE #18: Custom emoticons
Defined on the server. This will require uploading images - ping @eq since you've done a bunch of image uploading service work before! :tada:

~~No 50-emote limit, but maybe there should still be *some* limit -- does 500 seem reasonable? :)~~

### PR DESCRIPTION
**Pull request by heyitsmeuralex**
image

image

Resolves #18

### CODE REVIEW

**Comment by towerofnix** [comment]
Blocked by #133.

File: site/src/app.js [hunk 1]
```diff
@@ -43,7 +43,10 @@ app.use((state, emitter) => {
     state.secure = false
     state.ws = new util.WS(state.params.host)
 
-    state.ws.on('open', () => state.secure = state.ws.secure)
+    state.ws.on('open', () => {
+      state.secure = state.ws.secure
+      emitter.emit('emotes.fetch')
```
Comment (heyitsmeuralex) [inline site/src/app.js]: Note that this should probably be auto-fetched periodically (or when a new `emotes update` event is sent over the WS) on clients. At the moment you need to reload in order to see newly-added emoji.

File: site/src/components/messages.js [hunk 2]
```diff
@@ -266,6 +268,13 @@ const component = (state, emit) => {
     }
   }
 
+  if (state.messages.isScrolledToBottom()) {
+    // scroll to bottom after re-render
+    setTimeout(() => {
+      state.messages.el.scrollTop = state.messages.el.scrollHeight + 999
+    }, 50)
+  }
+
```
Comment (towerofnix) [inline site/src/components/messages.js]: Why the `+ 999`? Shouldn't `scrollHeight` automatically adjust enough?

Comment (heyitsmeuralex) [reply site/src/components/messages.js]: Yep, but it didn't.

JS is bad.

**Pull request merged**